lib/hotel.rb

```diff
@@ -0,0 +1,165 @@
+require_relative 'check_user_input'
+
+module BookingSystem
+  class Hotel
+    include BookingSystem::CheckUserInput
+
+    class UnavailableRoomError < StandardError
+    end
+    class UnavailableBlockError < StandardError
+    end
+
+    DEFAULT_ROOM_PRICE = 200
+
+    attr_reader :rooms, :all_single_reservations, :all_block_reservations, :all_blocks
+
+    def initialize
+      @rooms = (1..20).to_a
+      @all_single_reservations = []
+      @all_block_reservations = []
+      @all_blocks = []
+    end
+
+    def make_single_reservation(room, check_in, check_out, price = DEFAULT_ROOM_PRICE)
+      room_available = room_available?(room, check_in, check_out)
+
+      CheckUserInput.price(price)
+
+      if room_available == true
+        reservation = BookingSystem::Reservation.new(room, check_in, check_out, price)
+        @all_single_reservations << reservation
+
+        return reservation
+      else #false
+        raise UnavailableRoomError.new("Room #{room} is not available for those dates - #{check_in.to_s} to #{check_out.to_s}")
+      end
+    end
+
+    def block_reservation(room, check_in, check_out, price = DEFAULT_ROOM_PRICE)
+      CheckUserInput.integer(room)
+      CheckUserInput.between_1_20(room)
+      CheckUserInput.price(price)
+
+      requested_date_range = DateRange.new(check_in, check_out).all_reservation_dates
+
+      #check that there are any blocks
+      if @all_blocks.empty?
+        raise UnavailableBlockError.new("There are no blocks currently in the system")
+      end
+
+      #check that there is a matching block for the date range requested
+      blocks_dates_match = @all_blocks.select {|block| block.date_range == requested_date_range}
+      if blocks_dates_match.empty?
+        raise UnavailableBlockError.new("There is no block matching this date range: #{check_in} - #{check_out}")
+      end
+
+      #check that room requested is available in block
+      new_block_reservation = nil
+      blocks_dates_match.each do |block|
+        #if a match make block reservation
+        #remove room from availablity in block
+        if block.rooms_available.include?(room)
+          new_block_reservation = BookingSystem::BlockReservation.new(room, check_in, check_out, price)
+          block.remove_reserved_room_from_availability(room)
```

Comment:
I like that you call a method on `Block` on line 63 rather than modifying the block's data directly. This is a good example of loose coupling.